posthog/caching/calculate_results.py

```diff
@@ -138,8 +138,7 @@ def calculate_for_query_based_insight(
         else ExecutionMode.CACHE_ONLY_NEVER_CALCULATE,
     )
 
-    if "results" not in response:
-        # Translating `CacheMissResponse` to legacy insights shape
+    if response.keys() == {"cache_key"}:  # Translating `CacheMissResponse` to legacy insights shape
```

Comment:
This feels a little confusing and too knowledgable about the internal structure of CacheMissResponse.

Can we use `isinstance(response, CacheMissResponse)` or something else that doesn't assume internals?

Reply:
Because of pydantic, the code isn't beautiful but it would be like this
```
    try:
        CacheMissResponse(**response)
        return NothingInCacheResult(cache_key=response.get("cache_key"))
    except ValidationError as e:
        return InsightResult(
            # Translating `QueryResponse` to legacy insights shape
            # Only `results` is guaranteed even for non-insight queries, such as `EventsQueryResponse`
            result=response["results"],
            columns=response.get("columns"),
            last_refresh=response.get("last_refresh"),
            cache_key=response.get("cache_key"),
            is_cached=response.get("is_cached", False),
            timezone=response.get("timezone"),
            next_allowed_client_refresh=response.get("next_allowed_client_refresh"),
            timings=response.get("timings"),
        )
```

Reply:
It's annoying that `DatabaseSchemaQuery`, `TimeToSeeDataSessionsQuery`, and `SessionEventsQuerySerializer` responses are dicts, so `process_query()` serializes has to serialize _everything_ to dicts for consistency, causing us to lose out on Pydantic in this layer

Reply:
I like your suggestion in terms of being Pythonic, but I feel like in practice the try:catch: may be even less clear, hence I'm tempted to stick with the dirty `.keys()` check 😅

Reply:
Yeah i agree the try except is ugly and unclear.

I guess at least add a comment in schema.ts in CacheMissResponse that we do this, in case somebody tries to change it and it breaks.

It doesn't feel good though.

Reply:
Got rid of the unclarity around this by refactoring the processing functions to return without dict-ification